Added validation example for Problem 3.x of Class (2009) - Part 1 (#1853)
GEOSX/GEOSX: src/docs/sphinx/advancedExamples/validationStudies/carbonStorage/isothermalHystInjection/Example.rst

  ../../../../../../inputFiles/compositionalMultiphaseFlow/benchmarks/Class09Pb3/class09_pb3_smoke.xml
  ../../../../../../inputFiles/compositionalMultiphaseFlow/benchmarks/Class09Pb3/class09_pb3_drainageOnly_iterative_base.xml

------------------------------------------------------------------------
Problem description
------------------------------------------------------------------------

The following text is adapted from the detailed description of the benchmark test case
presented in
`(Class et al., 2009) <https://link.springer.com/article/10.1007/s10596-009-9146-x>`__.

The setup is illustrated in the figure below.
The mesh can be found in `GEOSXDATA <https://github.com/GEOSX/GEOSXDATA>`__ and
was provided for the benchmark. It discretizes the widely-used `Johansen` reservoir,
which consists in a tilted reservoir with a main fault. Both porosity and permeability are
heterogeneous an given at vertices.

A single CO2 injection well is completed with perforations only in the bottom 50 m
of the reservoir.

The model domain has the following dimensions: 9600 x 8900 x [90-140] m.
It has a single well located at (5440,3300) m. The injection takes place for 25 years over 50 years

> simulated at the constant rate of 15 kg/s. An hydrostatic pressure gradient is imposed on the boundary cells

Reviewer: ```suggestion
simulated at the constant rate of 15 kg/s. A hydrostatic pressure gradient is imposed on the boundary faces
```

as well as a constant geothermal gradient of 0.03 K/m. The reservoir is initially at 100 C.

   .. _JohansenScenario: .. figure:: problem3.png
      :align: center
      :width: 500
      :figclass: align-center

The authors have used the following simplifying assumptions:

- The formation is isotropic.
- All processes are isothermal.
- The pressure conditions at the lateral boundaries are constant over time, and equal to the initial hydrostatic condition.

------------------------------------------------------------------
Mesh and element regions
------------------------------------------------------------------

The proposed conform discretization is
fully hexahedral. Cells on the border have been flagged by VTK a new *attribute* in order to apply specific

> boundary condition on them.

Author: No, though it was the opportunity to show how to do it :)

<xml extract>

A VTK filter ``PointToCell`` is used to map properties from vertices to cells, which by default build an uniform average of
values over the cell.

The structured mesh is generated using some helpers python script from the formatted Point/Cells list provided.
It contains ``ncells`` cells. It is then imported using ``meshImport``

.. literalinclude:: ../../../../../../inputFiles/compositionalMultiphaseFlow/benchmarks/Class09Pb3/class09_pb3_smoke.xml
    :language: xml
    :start-after: <!-- SPHINX_MESH -->
    :end-before: <!-- SPHINX_MESH_END -->

The central wellbore is discretized as well thanks to internal tools (see :ref:`co2_TutorialCO2FieldCase`).
It counts five segment with a perforation in each. It has its own region ``wellRegion`` and control labeled ``wellControls``
defined and detailed respectively in **ElementRegions** and **Solvers** (see below).

In the **ElementRegions** block,

.. literalinclude:: ../../../../../../inputFiles/compositionalMultiphaseFlow/benchmarks/Class09Pb3/class09_pb3_smoke.xml
    :language: xml
    :start-after: <!-- SPHINX_ER -->
    :end-before: <!-- SPHINX_ER_END -->

> two reservoir region are labeled, ``reservoir`` and ``boundaries``, associated with ``1_hexahedra`` and `2_hexahedara`

Author: Again, sorry.

respectively for the computation of gas density and viscosity and the brine density and viscosity, along with pressure, temperature, salinity
discretization of the parameter space. The external *tables/co2flash.txt* gives the same type of information for the `CO2Solubility`
model. (see :ref:`CO2-EOS` for details).

The rock model defines a slightly compressible porous medium with a reference porosity equal to 0.1.

.. literalinclude:: ../../../../../../inputFiles/compositionalMultiphaseFlow/benchmarks/Class09Pb3/class09_pb3_drainageOnly_iterative_base.xml
    :language: xml
    :start-after: <!-- SPHINX_ROCK -->
    :end-before: <!-- SPHINX_ROCK_END -->

.. note::
    comment on NullSolid

The relative permeability model is input through tables thanks to **TableRelativePermeability** block.

.. literalinclude:: ../../../../../../inputFiles/compositionalMultiphaseFlow/benchmarks/Class09Pb3/class09_pb3_drainageOnly_iterative_base.xml
    :language: xml
    :start-after: <!-- SPHINX_FLUID -->
    :end-before: <!-- SPHINX_FLUID_END -->

As this benchmark is testing the sensitivity of the plume dynamics to the relative permeability hysteresis modeling,

> in commented block the **TableRelativePermeabilityHysteresis** block sets up bounding curves for imbibition and drainage

Author: No worries,I will use *_smoke.xml* and *_drainageOnly_iteratire_base.xml* mainly and few places in *_hystRelperm_iteratire_base.xml* to allow easier maintenance.

    :language: xml
    :start-after: <!-- SPHINX_FLUID -->
    :end-before: <!-- SPHINX_FLUID_END -->

-----------------------------------------------------------
Initial and boundary conditions
-----------------------------------------------------------

The domain is initially saturated with brine with a hydrostatic pressure field.
This is specified using the **HydrostaticEquilibrium** XML tag in the **FieldSpecifications** block.
The datum pressure and elevation used below are defined in
`(Class et al., 2009) <https://link.springer.com/article/10.1007/s10596-009-9146-x>`__).

.. literalinclude:: ../../../../../../inputFiles/compositionalMultiphaseFlow/benchmarks/Class09Pb3/class09_pb3_drainageOnly_iterative_base.xml
    :language: xml
    :start-after: <!-- SPHINX_HYDROSTATIC -->
    :end-before: <!-- SPHINX_HYDROSTATIC_END -->

In the **Functions** block, the **TableFunction** s named ``initOilCompFracTable`` and ``initWaterCompFracTable``
define the brine-saturated initial state, while the **TableFunction** named ``initTempTable`` defines the

> homogeneous temperature field (temperature is not used in this benchmark).   

Reviewer: Maybe we should remove the part about temperature, because here the temperature is not exactly homogeneous. With the geothermal gradient that we are using, there should be at least 1.5 K of temperature difference between the bottom and the top of the model. It is not much, but still

> Since the fluid densities are constant for this simple incompressible fluid model, the same result
> could have been achieved using a simpler table in a **FieldSpecification** tag, as explained
> in :ref:`EquilibriumInitialCondition`.

Reviewer: ```suggestion
```
Same here, I would remove because there a mildly nonlinear dependence of brine density on pressure (small) and on temperature (larger), and here we consider a geothermal gradient.

I agree this is slightly inconsistent with the way boundary conditions are imposed...

The boundaries are set to have a constant 0.03 K/m temperature gradient as well as the hydrostatic pressure gradient.
We supplement that with water dominant content. Each block is linking a `fieldName` to a **TableFunction** tagged as
the value of `functionName`. In order to have those imposed on the boundary faces, we provide ``faceManager`` as `objectPath`.

.. literalinclude:: ../../../../../../inputFiles/compositionalMultiphaseFlow/benchmarks/Class09Pb3/class09_pb3_drainageOnly_iterative_base.xml
    :language: xml
    :start-after: <!-- SPHINX_BC -->
    :end-before: <!-- SPHINX_BC_END -->

---------------------------------
Outputing reservoir Statistics
---------------------------------

> In order to output partitioning of co2 mass, we use reservoir statistics. This is done by defining a `Task`, with

Reviewer: ```suggestion
In order to output the partitioning of the CO2 mass, we use reservoir statistics implemented in GEOSX. This is done by defining a `Task`, with
```

    compflowStatistics, reservoir: Temperature (min, average, max): 293.15, 371.199, 379.828 K
    compflowStatistics, reservoir: Total dynamic pore volume: 1.57994e+09 rm^3
    compflowStatistics, reservoir: Phase dynamic pore volumes: { 1.00239e+06, 1.57894e+09 } rm^3
    compflowStatistics, reservoir: Phase mass (including both trapped and non-trapped): { 7.24891e+08, 1.6245e+12 } kg
    compflowStatistics, reservoir: Trapped phase mass: { 5.3075e+08, 3.2511e+11 } kg
    compflowStatistics, reservoir: Dissolved component mass: { { 7.13794e+08, 4.53966e+06 }, { 1.6338e+08, 1.62433e+12 } } kg
    compflowStatistics: Max phase CFL number: 65.1284
    compflowStatistics: Max component CFL number: 2.32854

.. note::

    The log file mentioned above could be an explicit printout of the stdio from MPI launch or the autogenerated output from
    a SLURM job *slurm.out* or similar

--------------------------
Inspecting results
---------------------------------

We request VTK-format output files and use Paraview to visualize the results under the **Outputs** block.

The following figure shows the distribution of CO2 saturation thresholded above a significant value (here 0.001).

> The displayed cells are colored with respect to the co2 mass they contain. If the relative permeability for the gas phase
> drops below 10e-7, the cell is change into black

Reviewer: ```suggestion
The displayed cells are colored with respect to the CO2 mass they contain. If the relative permeability for the gas phase
drops below 10e-7, the cell is displayed in black.
```

.. _pb3_drainOnly_CO2saturation:
.. figure:: classp3_anim2.gif
   :align: center
   :width: 1000
   :figclass: align-center

   Plume of CO2 saturation for significant value where immobile co2 is colored in black.

> 
> We observe the importance of hysteresis modeling in co2 plume migration. Indeed, during the migration phase, the cells in
> the tail of the plume are reverting from drainage to imbibition and remaining co2 is trapped. This results in a lot slower migration
> and expansion of the plume.

Reviewer: ```suggestion

We observe the importance of hysteresis modeling in CO2 plume migration. Indeed, during the migration phase, the cells at
the tail of the plume are switching from drainage to imbibition and the residual co2 is trapped. This results in a slower migration
and expansion of the plume.
```

To validate the GEOSX results, we consider the metrics used in
`(Class et al., 2009) <https://link.springer.com/article/10.1007/s10596-009-9146-x>`__.

> The reporting values are the dissolved and gaseous co2 with respect to time using only drainage relative permeability
> and hysteretical relative permeabilities.

Reviewer: ```suggestion
The reporting values are the dissolved and gaseous CO2 with respect to time using only the drainage relative permeability and using 
hysteretic relative permeabilities.
```

.. _pb3_dissolvedMobile:
.. figure:: pb3_curves.png
   :align: center
   :width: 1000
   :figclass: align-center

   CO2 dissolved, gaseous and total mass with respect to time for (left) drainage only and (right) hysteresis relative permeabilities

> We can see that at the end of the injection period the mass of co2 in the gaseous phase stop increasing and start
> decreasing under dissolution.

Reviewer: ```suggestion
We can see that at the end of the injection period the mass of CO2 in the gaseous phase stops increasing and starts
decreasing due to dissolution of CO2 in the brine phase.
```
